src/librustc/ich/impls_ty.rs

```diff
@@ -902,13 +902,59 @@ for ty::TypeVariants<'gcx>
             TyForeign(def_id) => {
                 def_id.hash_stable(hcx, hasher);
             }
-            TyInfer(..) => {
-                bug!("ty::TypeVariants::hash_stable() - Unexpected variant {:?}.", *self)
+            TyInfer(infer_ty) => {
+                infer_ty.hash_stable(hcx, hasher);
             }
         }
     }
 }
 
+impl_stable_hash_for!(enum ty::InferTy {
+    TyVar(a),
+    IntVar(a),
+    FloatVar(a),
+    FreshTy(a),
+    FreshIntTy(a),
+    FreshFloatTy(a),
```

Comment:
The `Fresh*` variants are valid to hash?

Reply:
Yes.

Reply:
They are not confined to the local tcx:

https://github.com/rust-lang/rust/blob/3b6412b94324b10f698a18ea5766ef6ff8921ae8/src/librustc/ty/flags.rs#L112-L124

Reply:
It just happens that we don't ever put them in query keys right now (and probably never will, but whatever).

Reply:
(To be honest, even the `TyVar` and so forth would be ok to hash, it's just that it truly *ought* to be impossible... it would indicate some other problem if we did see them that would be worth knowing about.)